src/System.Private.CoreLib/shared/System/Threading/WaitHandle.cs

```diff
@@ -0,0 +1,436 @@
+            GC.SuppressFinalize(this);
+        }
+
+        public virtual bool WaitOne(int millisecondsTimeout)
+        {
+            if (millisecondsTimeout < -1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(millisecondsTimeout), SR.ArgumentOutOfRange_NeedNonNegOrNegative1);
+            }
+
+            return WaitOneNoCheck(millisecondsTimeout);
+        }
+
+        private bool WaitOneNoCheck(int millisecondsTimeout)
+        {
+            Debug.Assert(millisecondsTimeout >= -1);
+
+            // The field value is modifiable via the public <see cref="WaitHandle.SafeWaitHandle"/> property, save it locally
+            // to ensure that one instance is used in all places in this method
+            SafeWaitHandle waitHandle = _waitHandle;
+            if (waitHandle == null)
+            {
+                // Throw ObjectDisposedException for backward compatibility even though it is not be representative of the issue
+                throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+            }
+
+            bool success = false;
+            try
+            {
+                int waitResult;
+
+                waitHandle.DangerousAddRef(ref success);
+
+                SynchronizationContext context = SynchronizationContext.Current;
+                if (context != null && context.IsWaitNotificationRequired())
+                {
+                    waitResult = context.Wait(new[] { waitHandle.DangerousGetHandle() }, false, millisecondsTimeout);
+                }
+                else
+                {
+                    waitResult = WaitOneCore(waitHandle.DangerousGetHandle(), millisecondsTimeout);
+                }
+
+                if (waitResult == WaitAbandoned)
+                {
+                    throw new AbandonedMutexException();
+                }
+
+                return waitResult != WaitTimeout;
+            }
+            finally
+            {
+                if (success)
+                    waitHandle.DangerousRelease();
+            }
+        }
+
+        // Returns an array for storing SafeWaitHandles in WaitMultiple calls. The array
+        // is reused for subsequent calls to reduce GC pressure.
+        private static SafeWaitHandle[] RentSafeWaitHandleArray(int capacity)
+        {
+            SafeWaitHandle[] safeWaitHandles = _safeWaitHandlesForRent;
+
+            // _safeWaitHandlesForRent can be null when it was not initialized yet or
+            // if a re-entrant wait is performed and the array is already rented. In
+            // that case we just allocate a new one and reuse it as necessary.
+            if (safeWaitHandles == null || safeWaitHandles.Length < capacity)
+            {
+                // Always allocate at least 4 slots to prevent unnecessary reallocations
+                safeWaitHandles = new SafeWaitHandle[Math.Max(capacity, 4)];
+            }
+
+            _safeWaitHandlesForRent = null;
```

Comment:
Nit: It would be a bit better to clear this just after `SafeWaitHandle[] safeWaitHandles = _safeWaitHandlesForRent;`. It is better to do thread local access close to each other - it helps the JIT to CSE pointer to the thread local block.